polygeist/lib/Dialect/Polygeist/Analysis/MemoryAccessAnalysis.cpp

```diff
@@ -641,6 +676,277 @@ MemoryAccess<OpTy>::classifyMemoryAccess(DataFlowSolver &solver) const {
+  return (index && index.getValue() == 0);
+}
+
+std::optional<Definition>
+MemoryAccessAnalysis::getUniqueDefinition(unsigned opIndex, Operation *op,
+                                          DataFlowSolver &solver) const {
+  using ModifiersTy = ReachingDefinition::ModifiersTy;
+
+  const ReachingDefinition *reachingDef =
+      solver.lookupState<ReachingDefinition>(op);
+  if (!reachingDef)
+    return std::nullopt;
+
+  Value operand = op->getOperand(opIndex);
+  std::optional<ModifiersTy> mods = reachingDef->getModifiers(operand);
+  std::optional<ModifiersTy> pMods =
+      reachingDef->getPotentialModifiers(operand);
+
+  // If there are potential modifiers then there is no unique modifier.
+  if (pMods.has_value() && !pMods->empty())
+    return std::nullopt;
+
+  if (!mods.has_value() || mods->empty())
+    return std::nullopt;
+
+  return *mods->begin();
+}
+
+SmallVector<Value>
+MemoryAccessAnalysis::getUnderlyingValues(unsigned opIndex, Operation *op,
+                                          DataFlowSolver &solver) const {
+  Value operand = op->getOperand(opIndex);
+  std::optional<Definition> def = getUniqueDefinition(opIndex, op, solver);
+  if (!def.has_value())
+    return {operand};
+
+  LLVM_DEBUG({
+    llvm::dbgs() << "operand: " << operand << "\n";
+    llvm::dbgs() << "operand definition: " << *def << "\n";
+  });
+
+  return TypeSwitch<Operation *, SmallVector<Value>>(def->getOperation())
+      .Case([&](AffineStoreOp storeOp) {
+        // Memory accesses involving SYCL accessors should have zero index.
+        MemRefAccess storeAccess(storeOp);
+        assert(hasZeroIndex(storeAccess) && "Unexpected candidate operation");
+
+        Value storedVal =
+            storeOp.getOperand(storeOp.getStoredValOperandIndex());
+        if (!isa<AffineLoadOp>(storedVal.getDefiningOp()))
+          return getUnderlyingValues(storeOp.getStoredValOperandIndex(),
+                                     storeOp, solver);
+
+        // Try to determine the underlying value of the memory pointed to by the
+        // memref operand of a load.
+        AffineLoadOp loadOp = cast<AffineLoadOp>(storedVal.getDefiningOp());
+        MemRefAccess loadAccess(loadOp);
+        assert(hasZeroIndex(storeAccess) && "Unexpected candidate operation");
+
+        return getUnderlyingValues(loadOp.getMemRefOperandIndex(), loadOp,
+                                   solver);
+      })
+      .Case([&](sycl::SYCLConstructorOp constructorOp) {
+        // Collect the underlying values of the sycl.constructor inputs.
+        SmallVector<Value> vec;
+        for (unsigned i = 1; i < constructorOp.getNumOperands(); ++i) {
+          auto vals = getUnderlyingValues(i, constructorOp, solver);
+          assert(vals.size() == 1 && "Expecting single value");
+          vec.push_back(vals.front());
+        }
+        return vec;
+      })
+      .Default([&](auto *op) {
+        SmallVector<Value> vec;
+        for (auto res : op->getResults())
```

Comment:
why should be the results?
thinking to just return empty vector for unexpected cases.